multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,327 @@
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
+		Complete(r)
+}
+
+// namespaceMapFunc handles Namespace update events (Namespace label change) by enqueuing
+// all Pods in the Namespace into the reconciler processing queue.
+func (r *LabelIdentityReconciler) namespaceMapFunc(ns client.Object) []reconcile.Request {
+	podList := &v1.PodList{}
+	r.Client.List(context.TODO(), podList, client.InNamespace(ns.GetName()))
+	requests := make([]reconcile.Request, len(podList.Items))
+	for i, pod := range podList.Items {
+		requests[i] = reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      pod.GetName(),
+				Namespace: pod.GetNamespace(),
+			},
+		}
+	}
+	return requests
+}
+
+// onPodDelete removes the Pod and label identity mapping from the cache, and deletes label identity
+// ResourceExport if necessary (the Pod deletion event causes a label identity no longer present in
+// the cluster).
+// Note that if releaseLabelMutex is set to false, r.labelMutex will continue to be held after the
+// function returns, which is useful for eliminating additional context switching when handling Pod
+// update events, since after deleting stale label the labelMutex will immediately need to be held
+// again for checking label add.
+func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podNamespacedName string, releaseLabelMutex bool) (ctrl.Result, error) {
+	r.labelMutex.Lock()
+	if originalLabel, isCached := r.podLabelCache[podNamespacedName]; isCached {
+		if requeue := r.requeueIfLabelResExportUpdateInProgress(originalLabel); requeue {
+			r.labelMutex.Unlock()
+			return ctrl.Result{Requeue: true}, nil
+		}
+		// Check if the original label is stale.
+		if podNames, ok := r.labelToPodsCache[originalLabel]; ok && len(podNames) == 1 && podNames.Has(podNamespacedName) {
+			klog.V(2).InfoS("Label no longer exists in the cluster, deleting corresponding ResourceExport", "label", originalLabel)
+			if err := r.deleteLabelIdentityResExport(ctx, originalLabel); err != nil {
+				r.labelMutex.Unlock()
+				return ctrl.Result{}, err
+			}
+			r.deleteFromLabelCache(podNamespacedName, originalLabel)
+		} else {
+			podNames.Delete(podNamespacedName)
+			delete(r.podLabelCache, podNamespacedName)
+		}
+	}
+	// Keep labelMutex locked if label add check will immediately follow.
+	if releaseLabelMutex {
+		r.labelMutex.Unlock()
+	}
+	return ctrl.Result{}, nil
+}
+
+// onPodCreateOrUpdate updates the Pod and label identity mapping in the cache, and
+// updates label identity ResourceExport if necessary (the Pod creation/update
+// event causes a new label identity to appear in the cluster or a label identity
+// no longer present in the cluster or both).
+func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podNamespacedName, currentLabel string) (ctrl.Result, error) {
+	if result, err := r.onPodDelete(ctx, podNamespacedName, false); err != nil || result.Requeue {
```

Comment:
We should not go to the delete processing, if the Pod's label does not change?

Reply:
Based on the reconciler's predicate it should not be possible. Worst case scenario if it happens, we would delete a ResourceExport and create a new one, which is harmless